File: 0979-distribute-coins-in-binary-tree/0979-distribute-coins-in-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def distributeCoins(self, root: Optional[TreeNode]) -> int:
        self.output = 0
        self.dfs(root, None)

        return self.output
        
    def dfs(self, current, parent):
        if current == None:
            return;
        
        self.dfs(current.left, current)
        self.dfs(current.right, current)

        if current.val < 1:
            self.output += 1 - current.val
            parent.val -= 1 - current.val
        elif current.val > 1:
            self.output += current.val - 1
            parent.val += current.val - 1

File: 0979-distribute-coins-in-binary-tree/0979-distribute-coins-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int output = 0;
    public int distributeCoins(TreeNode root) {
        dfs(root, null);

        return output;
    }

    public void dfs(TreeNode root, TreeNode parent) {
        if (root == null) return;

        System.out.printf("root.val: %d\n", root.val);

        dfs(root.left, root);
        dfs(root.right, root);
        
        if (root.val < 1) {
            System.out.printf("root.val < 1\n");
            System.out.printf("root.val: %d\n", root.val);
            System.out.printf("parent.val: %d\n", parent.val);
            output += (1 - root.val);
            parent.val -= (1 - root.val); 
            System.out.printf("After\n");
            System.out.printf("root.val: %d\n", root.val);
            System.out.printf("parent.val: %d\n", parent.val);
        }
        else if(root.val > 1) {
            System.out.printf("root.val > 1\n");
            System.out.printf("root.val: %d\n", root.val);
            System.out.printf("parent.val: %d\n", parent.val);
            output += root.val - 1;
            parent.val += root.val - 1; 
            System.out.printf("After\n");
            System.out.printf("root.val: %d\n", root.val);
            System.out.printf("parent.val: %d\n", parent.val);
        }
    }
}